pupil_src/shared_modules/pupil_detector_plugins/pye3d_plugin.py

```diff
@@ -0,0 +1,127 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2020 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+import logging
+
+from pye3d.detector_3d import Detector3D
+from pyglui import ui
+
+from .detector_base_plugin import PupilDetectorPlugin
+from .visualizer_2d import draw_eyeball_outline, draw_pupil_outline
+from .visualizer_pye3d import Eye_Visualizer
+
+logger = logging.getLogger(__name__)
+
+
+class Pye3DPlugin(PupilDetectorPlugin):
+    uniqueness = "by_class"
+    icon_font = "pupil_icons"
+    icon_chr = chr(0xEC19)
+
+    label = "Pye3D"
+    identifier = "3d"
+    order = 0.101
+
+    def __init__(self, g_pool):
+        super().__init__(g_pool)
+        self.detector = Detector3D()
+        self.debugVisualizer3D = Eye_Visualizer(
+            g_pool, self.detector.settings["focal_length"]
+        )
+
+    def detect(self, frame, **kwargs):
+        previous_detection_results = kwargs.get("previous_detection_results", [])
+        for datum in previous_detection_results:
+            if datum.get("method", "") == "2d c++":
+                datum_2d = datum
+                break
+        else:
+            # TODO: make this more stable!
+            raise RuntimeError("No 2D detection result! Needed for pye3D!")
+
+        datum_2d["raw_edges"] = []
+        result = self.detector.update_and_detect(
+            datum_2d, frame.gray, debug_toggle=self.is_debug_window_open
+        )
+
+        eye_id = self.g_pool.eye_id
+        result["timestamp"] = frame.timestamp
+        result["topic"] = f"pupil.{eye_id}.{self.identifier}"
+        result["id"] = eye_id
+        result["method"] = "3d c++"
```

Comment:
We might want to change that line though.
```suggestion
        result["method"] = "pye3d"
```

Reply:
Not sure if this would affect hmd-eyes though!

Reply:
This would be quite a big change actually.
Since Pupil v2.0 we adjusted the docs everywhere that you now get data from both detectors and that you should filter on the `method` key, e.g. for `3d c++`.
So a lot of users might rely on that and we would have to change a lot of documentation.
On the one hand, it makes sense that this is now a new method.
On the other hand, since we replace the old 3D we might just re-use the method key?

Reply:
I was not aware of the recommendation to filter for the complete string. The gazers only filter for the "2d"/"3d" part. In this case, I would resolve this with a `won't fix`